File: app/src/main/java/com/qwwuyu/file/helper/FileHelper.java
package com.qwwuyu.file.helper;

import android.os.Environment;

import com.qwwuyu.file.WApplication;
import com.qwwuyu.file.config.Constant;
import com.qwwuyu.file.config.ManageConfig;
import com.qwwuyu.file.entity.FileBean;
import com.qwwuyu.file.entity.ResponseBean;
import com.qwwuyu.file.utils.AppUtils;
import com.qwwuyu.file.utils.CommUtils;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * 文件工具类
 */
public class FileHelper {
    /** 单例 */
    private static FileHelper instance = new FileHelper();
    /** 是否创建路径成功 */
    private boolean isCreate = false;
    /** 临时存储路径 */
    private String cachePath;

    public static FileHelper getInstance() {
        return instance;
    }

    private FileHelper() {
        create();
    }

    /** 创建文件SD卡路径 */
    private void create() {
        if (CommUtils.isExternalEnable(WApplication.context)) {
            String externalPath = WApplication.context.getExternalCacheDir().toString();
            cachePath = externalPath + File.separator + "ManageCache" + File.separator;
            File file = new File(cachePath);
            if (file.isFile()) file.delete();
            isCreate = file.exists() || file.mkdirs();
            File[] files = file.listFiles();
            if (files != null) {
                for (File f : files) {
                    f.delete();
                }
            }
        }
    }

    public boolean checkCreate() {
        if (isCreate) return true;
        create();
        return isCreate;
    }

    public String getCachePath() {
        return cachePath;
    }

    public static List<FileBean> getDirectoryFile(String path) {
        File f = file(path);
        List<FileBean> fileBeans = new ArrayList<>();
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
        File[] files = f.listFiles();
        if (files != null) {
            for (File file : files) {
                String name = file.getName();
                if (ManageConfig.Companion.getInstance().isShowPointFile() || !name.startsWith(".")) {
                    fileBeans.add(new FileBean(name, file.isDirectory(), format.format(new Date(file.lastModified()))));
                }
            }
        }
        AppUtils.sortFileBean(fileBeans);
        return fileBeans;
    }

    public static ResponseBean delFile(String path) {
        File f = file(path);
        ResponseBean bean = new ResponseBean();
        if (!f.isFile()) {
            bean.setState(Constant.HTTP_ERR);
            bean.setInfo("文件不存在");
        } else if (f.delete()) {
            bean.setState(Constant.HTTP_SUC);
            bean.setInfo("删除成功");
        } else {
            bean.setState(Constant.HTTP_ERR);
            bean.setInfo("删除失败");
        }
        return bean;
    }

    public static ResponseBean delDir(String path) {
        File f = file(path);
        ResponseBean bean = new ResponseBean();
        bean.setState(Constant.HTTP_ERR);
        if (!f.isDirectory()) {
            bean.setInfo("文件夹不存在");
        } else {
            String[] list = f.list();
            if (list != null && list.length > 0) {
                bean.setInfo("文件夹包含其他文件，无法删除");
            } else if (f.delete()) {
                bean.setState(Constant.HTTP_SUC);
                bean.setInfo("删除成功");
            } else {
                bean.setInfo("删除失败");
            }
        }
        return bean;
    }

    public static ResponseBean createDir(String path, String dirName) {
        ResponseBean bean = new ResponseBean();
        bean.setState(Constant.HTTP_ERR);
        if (dirName == null || dirName.matches(".*[\\\\/:*?\"<>|]+.*")) {
            bean.setInfo("不能包含特殊字符\\/:*?\"<>|");
            return bean;
        }
        File file = file(path);
        if (!file.isDirectory()) {
            bean.setInfo("文件夹不存在");
            return bean;
        }
        File dirFile = new File(file, dirName);
        if (dirFile.exists()) {
            bean.setInfo("文件夹已存在");
            return bean;
        }
        if (dirFile.mkdir()) {
            bean.setState(Constant.HTTP_SUC);
            bean.setInfo("创建文件夹成功");
        } else {
            bean.setInfo("创建文件夹失败");
        }
        return bean;
    }

    public static File file(String path) {
        return new File(Environment.getExternalStorageDirectory().getAbsolutePath() + path);
    }
}

File: app/src/main/java/com/qwwuyu/file/helper/FileHelper.kt
package com.qwwuyu.file.helper

import android.os.Environment
import androidx.documentfile.provider.DocumentFile
import com.qwwuyu.file.WApplication
import com.qwwuyu.file.config.Constant
import com.qwwuyu.file.config.ManageConfig
import com.qwwuyu.file.entity.FileBean
import com.qwwuyu.file.entity.ResponseBean
import com.qwwuyu.file.nano.DocFile
import com.qwwuyu.file.nano.ProxyFile
import com.qwwuyu.file.nano.RealFile
import com.qwwuyu.file.utils.CommUtils
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * 文件工具类
 */
class FileHelper private constructor() {
    init {
        create()
    }

    /** 是否创建路径成功  */
    private var isCreate = false

    /** 临时存储路径  */
    var cachePath: String? = null
        private set

    /** 创建文件SD卡路径  */
    private fun create() {
        var cacheDir: File? = WApplication.context.cacheDir
        if (cacheDir == null && CommUtils.isExternalCacheEnable(WApplication.context)) {
            cacheDir = WApplication.context.externalCacheDir
        }
        cachePath = cacheDir.toString() + File.separator + "ManageCache" + File.separator
        val file = File(cachePath!!)
        if (file.isFile) file.delete()
        isCreate = file.isDirectory || file.mkdirs()
        file.listFiles()?.forEach { it.delete() }
    }

    fun checkCreate(): Boolean {
        if (!isCreate) create()
        return isCreate
    }

    companion object {
        /** 单例  */
        @JvmStatic
        val instance = FileHelper()

        @JvmStatic
        fun getFile(path: String): ProxyFile {
            val file = File(Environment.getExternalStorageDirectory().absolutePath + path)
            if (file.exists() && file.canWrite()) {
                return RealFile(file)
            }
            var df: DocumentFile? = RFileHelper.df ?: return RealFile(file)
            if (!path.startsWith("/Android/data")) return RealFile(file)
            val dirs = path.split("/").drop(3)
            for (dir in dirs) {
                df = df?.let { check(it, dir) }
                if (df == null) break
            }
            return DocFile(df ?: return RealFile(file))
        }

        @JvmStatic
        fun getDirectoryFile(file: ProxyFile?): List<FileBean> {
            val fileBeans: MutableList<FileBean> = ArrayList()
            val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault())
            (file ?: return fileBeans).listFiles().forEach { cf ->
                val name = cf.getName()
                if (ManageConfig.instance.isShowPointFile() || !name.startsWith(".")) {
                    val date: String = dateFormat.format(Date(cf.lastModified()))
                    val directory = cf.isDirectory()
                    val child = if (directory) cf.listFiles().size else 0
                    val info = if (directory) "${child}项" else CommUtils.getFileSize(cf.length())
                    val isInfo = !directory || ManageConfig.instance.isDirInfo()
                    val fileBean = FileBean(name, directory,
                        if (isInfo) date else null, if (isInfo) info else null)
                    fileBean.apk = ManageConfig.instance.isShowApk() && name.endsWith(".apk")
                    fileBeans.add(fileBean)
                }
            }
            fileBeans.sortWith { lhs, rhs ->
                if (lhs.dir && !rhs.dir) -1
                else if (!lhs.dir && rhs.dir) 1
                else lhs.name.compareTo(rhs.name)
            }
            return fileBeans
        }

        @JvmStatic
        fun delFile(file: ProxyFile?): ResponseBean {
            val bean = ResponseBean()
            if (file == null || !file.exists()) {
                bean.state = Constant.HTTP_ERR
                bean.info = "文件不存在"
            } else if (file.delete()) {
                bean.state = Constant.HTTP_SUC
                bean.info = "删除成功"
            } else {
                bean.state = Constant.HTTP_ERR
                bean.info = "删除失败"
            }
            return bean
        }

        @JvmStatic
        fun delDir(file: ProxyFile?): ResponseBean {
            val bean = ResponseBean()
            bean.state = Constant.HTTP_ERR
            if (file == null || !file.isDirectory()) {
                bean.info = "文件夹不存在"
            } else if (file.listFiles().isNotEmpty()) {
                bean.info = "文件夹包含其他文件，无法删除"
            } else if (file.delete()) {
                bean.state = Constant.HTTP_SUC
                bean.info = "删除成功"
            } else {
                bean.info = "删除失败"
            }
            return bean
        }

        @JvmStatic
        fun createDir(file: ProxyFile?, dirName: String?): ResponseBean {
            val bean = ResponseBean()
            bean.state = Constant.HTTP_ERR
            if (dirName == null || dirName.matches(".*[\\\\/:*?\"<>|]+.*".toRegex())) {
                bean.info = "不能包含特殊字符\\/:*?\"<>|"
                return bean
            }
            if (file == null || !file.isDirectory()) {
                bean.info = "文件夹不存在"
                return bean
            }
            val dirFile = file.child(dirName)
            if (dirFile != null && dirFile.exists()) {
                bean.info = "文件夹已存在"
                return bean
            }
            val create = file.createDirectory(dirName)
            if (create != null && create.exists()) {
                bean.state = Constant.HTTP_SUC
                bean.info = "创建文件夹成功"
            } else {
                bean.info = "创建文件夹失败"
            }
            return bean
        }

        private fun check(df: DocumentFile, dir: String): DocumentFile? {
            for (listFile in df.listFiles()) {
                if (listFile.isDirectory && listFile.name == dir) {
                    return listFile
                }
            }
            return null
        }
    }
}